drivers/usb/udc/udc_smartbond.c

```diff
@@ -0,0 +1,1764 @@
+#define EP0_FIFO_SIZE 8
+#define EP_FIFO_SIZE  64
+
+#define EP_MAX DT_INST_PROP(0, num_bidir_endpoints)
+_Static_assert(EP_MAX > 0 && EP_MAX <= 4, "num-bidir-endpoints out of range");
+
+/*
+ * DA146xx register fields and bit mask are very long. Filed masks repeat register names.
+ * Those convenience macros are a way to reduce complexity of register modification lines.
+ */
+#define GET_BIT(val, field)     (val & field##_Msk) >> field##_Pos
+#define REG_GET_BIT(reg, field) (USB->reg & USB_##reg##_##field##_Msk)
+#define REG_SET_BIT(reg, field) (USB->reg |= USB_##reg##_##field##_Msk)
+#define REG_CLR_BIT(reg, field) (USB->reg &= ~USB_##reg##_##field##_Msk)
+#define REG_SET_VAL(reg, field, val)                                                               \
+	(USB->reg = (USB->reg & ~USB_##reg##_##field##_Msk) | (val << USB_##reg##_##field##_Pos))
+
+struct udc_smartbond_config {
+	IRQn_Type udc_irq;
+	IRQn_Type vbus_irq;
+	uint8_t fifo_read_threshold;
+	uint16_t dma_min_transfer_size;
+};
+
+/* Node functional states */
+#define NFSR_NODE_RESET       0
+#define NFSR_NODE_RESUME      1
+#define NFSR_NODE_OPERATIONAL 2
+#define NFSR_NODE_SUSPEND     3
+/*
+ * Those two following states are added to allow going out of sleep mode
+ * using frame interrupt.  On remove wakeup RESUME state must be kept for
+ * at least 1ms. It is accomplished by using FRAME interrupt that goes
+ * through those two fake states before entering OPERATIONAL state.
+ */
+#define NFSR_NODE_WAKING      (0x10 | (NFSR_NODE_RESUME))
+#define NFSR_NODE_WAKING2     (0x20 | (NFSR_NODE_RESUME))
+
+struct smartbond_ep_reg_set {
+	volatile uint32_t epc_in;
+	volatile uint32_t txd;
+	volatile uint32_t txs;
+	volatile uint32_t txc;
+	volatile uint32_t epc_out;
+	volatile uint32_t rxd;
+	volatile uint32_t rxs;
+	volatile uint32_t rxc;
+};
+
+struct smartbond_ep_state {
+	struct udc_ep_config config;
+	struct smartbond_ep_reg_set *regs;
+	struct net_buf *buf;
+	/** Packet size sent or received so far. It is used to modify transferred field
+	 * after ACK is received or when filling ISO endpoint with size larger then
+	 * FIFO size.
+	 */
+	uint16_t last_packet_size;
+	uint8_t iso: 1; /** ISO endpoint */
+};
+
+static struct usb_smartbond_dma_cfg {
+	int tx_chan;
+	int rx_chan;
+	uint8_t tx_slot_mux;
+	uint8_t rx_slot_mux;
+	const struct device *tx_dev;
+	const struct device *rx_dev;
+	struct dma_config tx_cfg;
+	struct dma_config rx_cfg;
+	struct dma_block_config tx_block_cfg;
+	struct dma_block_config rx_block_cfg;
+} usbd_dma_cfg = {
+	.tx_chan = DT_DMAS_CELL_BY_NAME(DT_NODELABEL(usbd), tx, channel),
+	.tx_slot_mux = DT_DMAS_CELL_BY_NAME(DT_NODELABEL(usbd), tx, config),
+	.tx_dev = DEVICE_DT_GET(DT_DMAS_CTLR_BY_NAME(DT_NODELABEL(usbd), tx)),
+	.rx_chan = DT_DMAS_CELL_BY_NAME(DT_NODELABEL(usbd), rx, channel),
+	.rx_slot_mux = DT_DMAS_CELL_BY_NAME(DT_NODELABEL(usbd), rx, config),
+	.rx_dev = DEVICE_DT_GET(DT_DMAS_CTLR_BY_NAME(DT_NODELABEL(usbd), rx)),
+};
```

Comment:
Please move the instantiation  of usbd_dma_cfg to the bottom, either as part of  UDC_SMARTBOND_DEVICE_DEFINE(n) or above it.

Reply:
DMA configuration split into config (from DT) and data (that can change) and generated with along instance of device